README.md
# ESP32_C3_ISR_Servo Library

[![arduino-library-badge](https://www.ardu-badge.com/badge/ESP32_C3_ISR_Servo.svg?)](https://www.ardu-badge.com/ESP32_C3_ISR_Servo)
[![GitHub release](https://img.shields.io/github/release/khoih-prog/ESP32_C3_ISR_Servo.svg)](https://github.com/khoih-prog/ESP32_C3_ISR_Servo/releases)
[![GitHub](https://img.shields.io/github/license/mashape/apistatus.svg)](https://github.com/khoih-prog/ESP32_C3_ISR_Servo/blob/main/LICENSE)
[![contributions welcome](https://img.shields.io/badge/contributions-welcome-brightgreen.svg?style=flat)](#Contributing)
[![GitHub issues](https://img.shields.io/github/issues/khoih-prog/ESP32_C3_ISR_Servo.svg)](http://github.com/khoih-prog/ESP32_C3_ISR_Servo/issues)


<a href="https://www.buymeacoffee.com/khoihprog6" title="Donate to my libraries using BuyMeACoffee"><img src="https://cdn.buymeacoffee.com/buttons/v2/default-yellow.png" alt="Donate to my libraries using BuyMeACoffee" style="height: 50px !important;width: 181px !important;" ></a>
<a href="https://www.buymeacoffee.com/khoihprog6" title="Donate to my libraries using BuyMeACoffee"><img src="https://img.shields.io/badge/buy%20me%20a%20coffee-donate-orange.svg?logo=buy-me-a-coffee&logoColor=FFDD00" style="height: 20px !important;width: 200px !important;" ></a>


---
---

## Table of Contents

* [Why do we need this ESP32_C3_ISR_Servo library](#why-do-we-need-this-esp32_c3_isr_servo-library)
  * [Features](#features)
  * [Important Notes about using ISR](#important-notes-about-using-isr)
  * [Currently supported Boards](#currently-supported-boards)
* [Changelog](changelog.md)
* [Prerequisites](#prerequisites)
* [Installation](#installation)
  * [Use Arduino Library Manager](#use-arduino-library-manager)
  * [Manual Install](#manual-install)
  * [VS Code & PlatformIO](#vs-code--platformio)
* [HOWTO Fix `Multiple Definitions` Linker Error](#howto-fix-multiple-definitions-linker-error) 
* [HOWTO Use analogRead() with ESP32 running WiFi and/or BlueTooth (BT/BLE)](#howto-use-analogread-with-esp32-running-wifi-andor-bluetooth-btble)
  * [1. ESP32 has 2 ADCs, named ADC1 and ADC2](#1--esp32-has-2-adcs-named-adc1-and-adc2)
  * [2. ESP32 ADCs functions](#2-esp32-adcs-functions)
  * [3. ESP32 WiFi uses ADC2 for WiFi functions](#3-esp32-wifi-uses-adc2-for-wifi-functions)
* [More useful Information](#more-useful-information)
  * [ESP32 Hardware Timers](#esp32-hardware-timers)
  * [New functions](#new-functions)
  * [What special in this ESP32_C3_ISR_Servo library](#what-special-in-this-esp32_c3_isr_servo-library)
* [HOWTO Usage](#howto-usage)
* [Examples](#examples)
  * [ 1. ESP32_C3_ISR_MultiServos](examples/ESP32_C3_ISR_MultiServos)
  * [ 2. ESP32_C3_MultipleRandomServos](examples/ESP32_C3_MultipleRandomServos)
  * [ 3. ESP32_C3_MultipleServos](examples/ESP32_C3_MultipleServos)
  * [ 4. ISR_MultiServos](examples/ISR_MultiServos)
  * [ 5. MultipleRandomServos](examples/MultipleRandomServos)
  * [ 6. MultipleServos](examples/MultipleServos)
  * [ 7. **multiFileProject**](examples/multiFileProject) **New**
* [Example ESP32_C3_ISR_MultiServos](#example-esp32_c3_isr_multiservos)
  * [1. File ESP32_C3_ISR_MultiServos.ino](#1-file-esp32_c3_isr_multiservosino)
* [Debug Terminal Output Samples](#debug-terminal-output-samples)
  * [1. ESP32_C3_MultipleRandomServos on ESP32C3_DEV](#1-esp32_c3_multiplerandomservos-on-esp32c3_dev)
  * [2. ESP32_C3_ISR_MultiServos on ESP32C3_DEV](#2-esp32_c3_isr_multiservos-on-esp32c3_dev)
* [Debug](#debug)
* [Troubleshooting](#troubleshooting)
* [Releases](#releases)
* [Issues](#issues)
* [TO DO](#to-do)
* [DONE](#done)
* [Contributions and Thanks](#contributions-and-thanks)
* [Contributing](#contributing)
* [License](#license)
* [Copyright](#copyright)


---
---

### Why do we need this [ESP32_C3_ISR_Servo library](https://github.com/khoih-prog/ESP32_C3_ISR_Servo)

#### Features

Imagine you have a system with a **mission-critical function** controlling a **robot arm** or doing something much more important. You normally use a software timer to poll, or even place the function in `loop()`. But what if another function is blocking the `loop()` or `setup()`.

So your function might not be executed, and the result would be disastrous.

You'd prefer to have your function called, no matter what happening with other functions (busy loop, bug, etc.).

The correct choice is to use a **Hardware Timer with Interrupt** to call your function.

These hardware timers, using interrupt, still work even if other functions are blocking. Moreover, they are **much more precise** (certainly depending on clock frequency accuracy) than other software timers using `millis()` or `micros()`. That's necessary if you need to measure some data requiring better accuracy.

Functions using normal software timers, relying on `loop()` and calling `millis()`, won't work if the **loop() or setup() is blocked by certain operation**. For example, certain function is blocking while it's connecting to WiFi or some services.

This library enables you to use `1 Hardware Timer` on an `ESP32_C3-based` board to control up to `16 independent servo motors`.


#### Important Notes about using ISR

1. Inside the attached function, `delay()` won’t work and the value returned by `millis()` will not increment. Serial data received while in the function may be lost. You should declare as volatile any variables that you modify within the attached function.

2. Typically global variables are used to pass data between an ISR and the main program. To make sure variables shared between an ISR and the main program are updated correctly, declare them as volatile.

3. Avoid using `Serial.print()`-related functions inside ISR. Just for temporary debug purpose, but even this also can crash the system any time. Beware.

4. Your functions are now part of **ISR (Interrupt Service Routine)**, and must be `lean / mean`, and follow certain rules. More to read on:

[HOWTO Attach Interrupt](https://www.arduino.cc/reference/en/language/functions/external-interrupts/attachinterrupt/)


#### Currently supported Boards

This [**ESP32_C3_ISR_Servo** library](https://github.com/khoih-prog/ESP32_C3_ISR_Servo) currently supports these following boards:

 1. **ESP32_C3-based boards**.
 
---
---


## Prerequisites

1. [`Arduino IDE 1.8.19+` for Arduino](https://github.com/arduino/Arduino). [![GitHub release](https://img.shields.io/github/release/arduino/Arduino.svg)](https://github.com/arduino/Arduino/releases/latest)
2. [`ESP32 Core 2.0.5+`](https://github.com/espressif/arduino-esp32) for ESP32-based boards. [![Latest release](https://img.shields.io/github/release/espressif/arduino-esp32.svg)](https://github.com/espressif/arduino-esp32/releases/latest/)

---

## Installation

### Use Arduino Library Manager

The best and easiest way is to use `Arduino Library Manager`. Search for `ESP32_C3_ISR_Servo`, then select / install the latest version.
You can also use this link [![arduino-library-badge](https://www.ardu-badge.com/badge/ESP32_C3_ISR_Servo.svg?)](https://www.ardu-badge.com/ESP32_C3_ISR_Servo) for more detailed instructions.

### Manual Install

Another way to install is to:

1. Navigate to [ESP32_C3_ISR_Servo](https://github.com/khoih-prog/ESP32_C3_ISR_Servo) page.
2. Download the latest release `ESP32_C3_ISR_Servo-main.zip`.
3. Extract the zip file to `ESP32_C3_ISR_Servo-main` directory 
4. Copy whole `ESP32_C3_ISR_Servo-main` folder to Arduino libraries' directory such as `~/Arduino/libraries/`.

### VS Code & PlatformIO

1. Install [VS Code](https://code.visualstudio.com/)
2. Install [PlatformIO](https://platformio.org/platformio-ide)
3. Install [**ESP32_C3_ISR_Servo** library](https://registry.platformio.org/libraries/khoih-prog/ESP32_C3_ISR_Servo) by using [Library Manager](https://registry.platformio.org/libraries/khoih-prog/ESP32_C3_ISR_Servo/installation). Search for **ESP32_C3_ISR_Servo** in [Platform.io Author's Libraries](https://platformio.org/lib/search?query=author:%22Khoi%20Hoang%22)
4. Use included [platformio.ini](platformio/platformio.ini) file from examples to ensure that all dependent libraries will installed automatically. Please visit documentation for the other options and examples at [Project Configuration File](https://docs.platformio.org/page/projectconf.html)


---
---

### HOWTO Fix `Multiple Definitions` Linker Error

The current library implementation, using `xyz-Impl.h` instead of standard `xyz.cpp`, possibly creates certain `Multiple Definitions` Linker error in certain use cases.

You can include this `.hpp` file

```cpp
// Can be included as many times as necessary, without `Multiple Definitions` Linker Error
#include "ESP32_C3_ISR_Servo.hpp"     //https://github.com/khoih-prog/ESP32_C3_ISR_Servo
```

in many files. But be sure to use the following `.h` file **in just 1 `.h`, `.cpp` or `.ino` file**, which must **not be included in any other file**, to avoid `Multiple Definitions` Linker Error

```cpp
// To be included only in main(), .ino with setup() to avoid `Multiple Definitions` Linker Error
#include "ESP32_C3_ISR_Servo.h"           //https://github.com/khoih-prog/ESP32_C3_ISR_Servo
```

Check the new [**multiFileProject** example](examples/multiFileProject) for a `HOWTO` demo.

Have a look at the discussion in [Different behaviour using the src_cpp or src_h lib #80](https://github.com/khoih-prog/ESPAsync_WiFiManager/discussions/80)


---
---

### HOWTO Use analogRead() with ESP32 running WiFi and/or BlueTooth (BT/BLE)

Please have a look at [**ESP_WiFiManager Issue 39: Not able to read analog port when using the autoconnect example**](https://github.com/khoih-prog/ESP_WiFiManager/issues/39) to have more detailed description and solution of the issue.

#### 1.  ESP32 has 2 ADCs, named ADC1 and ADC2

#### 2. ESP32 ADCs functions

- `ADC1` controls ADC function for pins **GPIO32-GPIO39**
- `ADC2` controls ADC function for pins **GPIO0, 2, 4, 12-15, 25-27**

#### 3.. ESP32 WiFi uses ADC2 for WiFi functions

Look in file [**adc_common.c**](https://github.com/espressif/esp-idf/blob/master/components/driver/adc_common.c)

> In `ADC2`, there're two locks used for different cases:
> 1. lock shared with app and Wi-Fi:
>    ESP32:
>         When Wi-Fi using the `ADC2`, we assume it will never stop, so app checks the lock and returns immediately if failed.
>    ESP32S2:
>         The controller's control over the ADC is determined by the arbiter. There is no need to control by lock.
> 
> 2. lock shared between tasks:
>    when several tasks sharing the `ADC2`, we want to guarantee
>    all the requests will be handled.
>    Since conversions are short (about 31us), app returns the lock very soon,
>    we use a spinlock to stand there waiting to do conversions one by one.
> 
> adc2_spinlock should be acquired first, then adc2_wifi_lock or rtc_spinlock.


- In order to use `ADC2` for other functions, we have to **acquire complicated firmware locks and very difficult to do**
- So, it's not advisable to use `ADC2` with WiFi/BlueTooth (BT/BLE).
- Use `ADC1`, and pins `GPIO32-GPIO39`
- If somehow it's a must to use those pins serviced by `ADC2` (**GPIO0, 2, 4, 12, 13, 14, 15, 25, 26 and 27**), use the **fix mentioned at the end** of [**ESP_WiFiManager Issue 39: Not able to read analog port when using the autoconnect example**](https://github.com/khoih-prog/ESP_WiFiManager/issues/39) to work with ESP32 WiFi/BlueTooth (BT/BLE).

---
---

## More useful Information

### ESP32 Hardware Timers

  - **The ESP32 has two timer groups, each one with two general purpose hardware timers.**
  - All the timers are based on **64-bit counters and 16-bit prescalers.**
  - The timer counters can be configured to count up or down and support automatic reload and software reload.
  - They can also generate alarms when they reach a specific value, defined by the software. 
  - The value of the counter can be read by the software program.
  

### New functions

```cpp
// returns last position in degrees if success, or -1 on wrong servoIndex
int getPosition(unsigned servoIndex);

// returns pulseWidth in microsecs (within min/max range) if success, or 0 on wrong servoIndex
unsigned int getPulseWidth(unsigned servoIndex);
```

### What special in this [ESP32_C3_ISR_Servo library](https://github.com/khoih-prog/ESP32_C3_ISR_Servo)

Now these new **16 ISR-based Servo controllers** just use one ESP32 Hardware Timer. The number 16 is just arbitrarily chosen, and depending on application, you can increase that number to 32, 48, etc. without problem.

The accuracy is nearly perfect compared to software timers. The most important feature is they're ISR-based timers

Therefore, their executions are not blocked by bad-behaving functions / tasks. This important feature is absolutely necessary for mission-critical tasks. 

The [**MultipleServos**](examples/MultipleServos) example, which controls 6 servos independently, will demonstrate the nearly perfect accuracy.
Being ISR-based servo controllers, their executions are not blocked by bad-behaving functions / tasks, such as connecting to WiFi, Internet and Blynk services.

This non-being-blocked important feature is absolutely necessary for mission-critical tasks.

You'll see blynkTimer Software is blocked while system is connecting to WiFi / Internet / Blynk, as well as by blocking task in loop(), using delay() function as an example. The elapsed time then is very unaccurate

---
---

## HOWTO Usage

How to use:

```cpp
#if !defined( ARDUINO_ESP32C3_DEV )
	#error This code is intended to run on the ESP32_C3 platform! Please check your Tools->Board setting.
#endif

#define TIMER_INTERRUPT_DEBUG       1
#define ISR_SERVO_DEBUG             1

// Select different ESP32 timer number (0-1) to avoid conflict
#define USE_ESP32_TIMER_NO          1

#include "ESP32_C3_ISR_Servo.h"

// Don't use PIN_D1 in core v2.0.0 and v2.0.1. Check https://github.com/espressif/arduino-esp32/issues/5868
// Don't use PIN_D2 with ESP32_C3 (crash)

//See file .../hardware/espressif/esp32/variants/(esp32|doitESP32devkitV1)/pins_arduino.h
#define PIN_LED           2         // Pin D2 mapped to pin GPIO2/ADC12 of ESP32, control on-board LED

#define PIN_D0            0         // Pin D0 mapped to pin GPIO0/BOOT/ADC11/TOUCH1 of ESP32
#define PIN_D1            1         // Pin D1 mapped to pin GPIO1/TX0 of ESP32
#define PIN_D2            2         // Pin D2 mapped to pin GPIO2/ADC12/TOUCH2 of ESP32
#define PIN_D3            3         // Pin D3 mapped to pin GPIO3/RX0 of ESP32
#define PIN_D4            4         // Pin D4 mapped to pin GPIO4/ADC10/TOUCH0 of ESP32
#define PIN_D5            5         // Pin D5 mapped to pin GPIO5/SPISS/VSPI_SS of ESP32
#define PIN_D6            6         // Pin D6 mapped to pin GPIO6/FLASH_SCK of ESP32
#define PIN_D7            7         // Pin D7 mapped to pin GPIO7/FLASH_D0 of ESP32
#define PIN_D8            8         // Pin D8 mapped to pin GPIO8/FLASH_D1 of ESP32
#define PIN_D9            9         // Pin D9 mapped to pin GPIO9/FLASH_D2 of ESP32

// Published values for SG90 servos; adjust if needed
#define MIN_MICROS      800  //544
#define MAX_MICROS      2450

int servoIndex1  = -1;
int servoIndex2  = -1;

void setup()
{
	Serial.begin(115200);

	while (!Serial && millis() < 5000);

  delay(500);

	Serial.print(F("\nStarting ISR_MultiServos on "));
	Serial.println(ARDUINO_BOARD);
	Serial.println(ESP32_C3_ISR_SERVO_VERSION);

	//Select ESP32 timer USE_ESP32_TIMER_NO
	ESP32_ISR_Servos.useTimer(USE_ESP32_TIMER_NO);

	servoIndex1 = ESP32_ISR_Servos.setupServo(PIN_D5, MIN_MICROS, MAX_MICROS);
	servoIndex2 = ESP32_ISR_Servos.setupServo(PIN_D6, MIN_MICROS, MAX_MICROS);

	if (servoIndex1 != -1)
		Serial.println(F("Setup Servo1 OK"));
	else
		Serial.println(F("Setup Servo1 failed"));

	if (servoIndex2 != -1)
		Serial.println(F("Setup Servo2 OK"));
	else
		Serial.println(F("Setup Servo2 failed"));
}

void loop()
{
	int position;

	if ( ( servoIndex1 != -1) && ( servoIndex2 != -1) )
	{
		for (position = 0; position <= 180; position++)
		{
			// goes from 0 degrees to 180 degrees
			// in steps of 1 degree

			if (position % 30 == 0)
			{
				Serial.print(F("Servo1 pos = "));
				Serial.print(position);
				Serial.print(F(", Servo2 pos = "));
				Serial.println(180 - position);
			}

			ESP32_ISR_Servos.setPosition(servoIndex1, position);
			ESP32_ISR_Servos.setPosition(servoIndex2, 180 - position);
			// waits 30ms for the servo to reach the position
			delay(30);
		}

		delay(5000);

		for (position = 180; position >= 0; position--)
		{
			// goes from 180 degrees to 0 degrees
			if (position % 30 == 0)
			{
				Serial.print(F("Servo1 pos = "));
				Serial.print(position);
				Serial.print(F(", Servo2 pos = "));
				Serial.println(180 - position);
			}

			ESP32_ISR_Servos.setPosition(servoIndex1, position);
			ESP32_ISR_Servos.setPosition(servoIndex2, 180 - position);
			// waits 30ms for the servo to reach the position
			delay(30);
		}

		delay(5000);
	}
}
```

---
---

### Examples: 

 1. [ESP32_C3_ISR_MultiServos](examples/ESP32_C3_ISR_MultiServos)
 2. [ESP32_C3_MultipleRandomServos](examples/ESP32_C3_MultipleRandomServos) 
 3. [ESP32_C3_MultipleServos](examples/ESP32_C3_MultipleServos) 
 4. [ISR_MultiServos](examples/ISR_MultiServos)
 5. [MultipleRandomServos](examples/MultipleRandomServos)
 6. [MultipleServos](examples/MultipleServos)
 7. [multiFileProject](examples/multiFileProject) **New**
 
---

### Example [ESP32_C3_ISR_MultiServos](examples/ESP32_C3_ISR_MultiServos)


#### 1. File [ESP32_C3_ISR_MultiServos.ino](examples/ESP32_C3_ISR_MultiServos/ESP32_C3_ISR_MultiServos.ino)

https://github.com/khoih-prog/ESP32_C3_ISR_Servo/blob/f6a14a464742a98edda265ce332289a3a9b363eb/examples/ESP32_C3_ISR_MultiServos/ESP32_C3_ISR_MultiServos.ino#L72-L184

---
---

### Debug Terminal Output Samples

### 1. ESP32_C3_MultipleRandomServos on ESP32C3_DEV

```
Starting ESP32_C3_MultipleRandomServos on ESP32C3_DEV
ESP32_C3_ISR_Servo v1.2.0
[ISR_SERVO] ESP32_C3_FastTimerInterrupt: _timerNo = 1 , _fre = 1000000
[ISR_SERVO] TIMER_BASE_CLK = 80000000 , TIMER_DIVIDER = 80
[ISR_SERVO] Starting  ITimer OK
Setup OK Servo index = 0
Setup OK Servo index = 1

changelog.md
# ESP32_C3_ISR_Servo Library

[![arduino-library-badge](https://www.ardu-badge.com/badge/ESP32_C3_ISR_Servo.svg?)](https://www.ardu-badge.com/ESP32_C3_ISR_Servo)
[![GitHub release](https://img.shields.io/github/release/khoih-prog/ESP32_C3_ISR_Servo.svg)](https://github.com/khoih-prog/ESP32_C3_ISR_Servo/releases)
[![GitHub](https://img.shields.io/github/license/mashape/apistatus.svg)](https://github.com/khoih-prog/ESP32_C3_ISR_Servo/blob/main/LICENSE)
[![contributions welcome](https://img.shields.io/badge/contributions-welcome-brightgreen.svg?style=flat)](#Contributing)
[![GitHub issues](https://img.shields.io/github/issues/khoih-prog/ESP32_C3_ISR_Servo.svg)](http://github.com/khoih-prog/ESP32_C3_ISR_Servo/issues)


<a href="https://www.buymeacoffee.com/khoihprog6" title="Donate to my libraries using BuyMeACoffee"><img src="https://cdn.buymeacoffee.com/buttons/v2/default-yellow.png" alt="Donate to my libraries using BuyMeACoffee" style="height: 50px !important;width: 181px !important;" ></a>
<a href="https://www.buymeacoffee.com/khoihprog6" title="Donate to my libraries using BuyMeACoffee"><img src="https://img.shields.io/badge/buy%20me%20a%20coffee-donate-orange.svg?logo=buy-me-a-coffee&logoColor=FFDD00" style="height: 20px !important;width: 200px !important;" ></a>

---
---

## Table of Contents

* [Changelog](#changelog)
  * [Releases v1.2.0](#releases-v120)
  * [Releases v1.1.0](#releases-v110)
 
---
---

## Changelog

### Releases v1.2.0

1. Update to use with ESP32 core v2.0.5+. 
2. Convert to `h-only` style
3. Use `allman astyle` and add `utils`

### Releases v1.1.0

1. Basic 16 ISR-based servo controllers using 1 hardware timer for `ESP32_C3`
2. Tested with ESP32 core `v2.0.2+` and `v1.0.6`



src/ESP32_C3_FastTimerInterrupt.h
/****************************************************************************************************************************
  ESP32_C3_FastTimerInterrupt.h
  For ESP32_C3 boards
  Written by Khoi Hoang

  Built by Khoi Hoang https://github.com/khoih-prog/ESP32_C3_ISR_Servo
  Licensed under MIT license

  Now with these new 16 ISR-based timers, the maximum interval is practically unlimited (limited only by unsigned long miliseconds)
  The accuracy is nearly perfect compared to software timers. The most important feature is they're ISR-based timers
  Therefore, their executions are not blocked by bad-behaving functions / tasks.
  This important feature is absolutely necessary for mission-critical tasks.

  Loosely based on SimpleTimer - A timer library for Arduino.
  Author: mromani@ottotecnica.com
  Copyright (c) 2010 OTTOTECNICA Italy

  Based on BlynkTimer.h
  Author: Volodymyr Shymanskyy

  Version: 1.1.0

  Version Modified By  Date        Comments
  ------- -----------  ----------  -----------
  1.1.0   K Hoang      03/08/2021  Initial coding for ESP32_C3
 *****************************************************************************************************************************/

#pragma once

#ifndef ESP32_C3_FastTimerInterrupt_h
#define ESP32_C3_FastTimerInterrupt_h

#if !defined( ARDUINO_ESP32C3_DEV )
  #error This code is intended to run on the ESP32_C3 platform! Please check your Tools->Board setting.
#endif

#include "ESP32_C3_ISR_Servo_Debug.h"

#include <driver/timer.h>

/*
  // ESP32 core v2.0.0-rc1, hw_timer_t defined in tools/sdk/esp32c3/include/hal/include/hal/timer_types.h:

  #define TIMER_BASE_CLK   (APB_CLK_FREQ)  //Frequency of the clock on the input of the timer groups 

  // In tools/sdk/esp32c3/include/soc/esp32c3/include/soc/soc_caps.h
  #define SOC_TIMER_GROUP_TIMERS_PER_GROUP  (1)
  #define SOC_TIMER_GROUPS                  (2)
  #define SOC_TIMER_GROUP_TOTAL_TIMERS      (SOC_TIMER_GROUPS * SOC_TIMER_GROUP_TIMERS_PER_GROUP)
  
 //@brief Selects a Timer-Group out of 2 available groups
 
typedef enum {
    TIMER_GROUP_0 = 0, //!<Hw timer group 0
#if SOC_TIMER_GROUPS > 1
    TIMER_GROUP_1 = 1, //!<Hw timer group 1
#endif
    TIMER_GROUP_MAX,
} timer_group_t;


 //@brief Select a hardware timer from timer groups
 
typedef enum {
    TIMER_0 = 0, //!<Select timer0 of GROUPx
#if SOC_TIMER_GROUP_TIMERS_PER_GROUP > 1
    TIMER_1 = 1, //!<Select timer1 of GROUPx
#endif
    TIMER_MAX,
} timer_idx_t;


 //@brief Decides the direction of counter
 
typedef enum 
{
  TIMER_COUNT_DOWN = 0, //Descending Count from cnt.high|cnt.low
  TIMER_COUNT_UP = 1,   //Ascending Count from Zero
  TIMER_COUNT_MAX
} timer_count_dir_t;


 //@brief Decides whether timer is on or paused
 
typedef enum 
{
  TIMER_PAUSE = 0,      //Pause timer counter
  TIMER_START = 1,      //Start timer counter
} timer_start_t;

  // @brief Interrupt types of the timer.
  //this is compatible with the value of esp32.
typedef enum 
{
    TIMER_INTR_T0 = BIT(0),  //!< interrupt of timer 0  <========== only this timer available for ESP32_C3
#if SOC_TIMER_GROUP_TIMERS_PER_GROUP > 1
    TIMER_INTR_T1 = BIT(1),  //!< interrupt of timer 1
    TIMER_INTR_WDT = BIT(2), //!< interrupt of watchdog
#else
    TIMER_INTR_WDT = BIT(1), //!< interrupt of watchdog
#endif
    TIMER_INTR_NONE = 0
} timer_intr_t;

 //@brief Decides whether to enable alarm mode
 
typedef enum 
{
  TIMER_ALARM_DIS = 0,  //Disable timer alarm
  TIMER_ALARM_EN = 1,   //Enable timer alarm
  TIMER_ALARM_MAX
} timer_alarm_t;


 //@brief Select interrupt type if running in alarm mode.
 
typedef enum 
{
  TIMER_INTR_LEVEL = 0,  //Interrupt mode: level mode
  TIMER_INTR_MAX
} timer_intr_mode_t;


 //@brief Select if Alarm needs to be loaded by software or automatically reload by hardware.
 
typedef enum 
{
  TIMER_AUTORELOAD_DIS = 0,  //Disable auto-reload: hardware will not load counter value after an alarm event
  TIMER_AUTORELOAD_EN = 1,   //Enable auto-reload: hardware will load counter value after an alarm event
  TIMER_AUTORELOAD_MAX,
} timer_autoreload_t;

#if SOC_TIMER_GROUP_SUPPORT_XTAL
// @brief Select timer source clock.
typedef enum 
{
    TIMER_SRC_CLK_APB = 0,  //!< Select APB as the source clock
    TIMER_SRC_CLK_XTAL = 1, //!< Select XTAL as the source clock
} timer_src_clk_t;
#endif

 //@brief Data structure with timer's configuration settings
 
typedef struct 
{
  timer_alarm_t alarm_en;           //!< Timer alarm enable
  timer_start_t counter_en;         //!< Counter enable
  timer_intr_mode_t intr_type;      //< Interrupt mode
  timer_count_dir_t counter_dir;    //!< Counter direction
  timer_autoreload_t auto_reload;   //!< Timer auto-reload
  uint32_t divider;   //!< Counter clock divider. The divider's range is from from 2 to 65536.
#if SOC_TIMER_GROUP_SUPPORT_XTAL
  timer_src_clk_t clk_src;          //!< Use XTAL as source clock.
#endif
} timer_config_t;

*/

#ifndef USE_ESP32_TIMER_NO
  #define USE_ESP32_TIMER_NO          3
#endif


class ESP32FastTimerInterrupt;

typedef ESP32FastTimerInterrupt ESP32FastTimer;

#define MAX_ESP32_NUM_TIMERS      4

#define TIMER_DIVIDER             80                                //  Hardware timer clock divider
// TIMER_BASE_CLK = APB_CLK_FREQ = Frequency of the clock on the input of the timer groups
#define TIMER_SCALE               (TIMER_BASE_CLK / TIMER_DIVIDER)  // convert counter value to seconds


// In esp32/2.0.0-rc1/tools/sdk/esp32c3/include/driver/include/driver/timer.h
// typedef bool (*timer_isr_t)(void *);
//esp_err_t timer_isr_callback_add(timer_group_t group_num, timer_idx_t timer_num, timer_isr_t isr_handler, void *arg, int intr_alloc_flags);
//esp_err_t timer_isr_callback_remove(timer_group_t group_num, timer_idx_t timer_num);
//timer_deinit(timer_group_t group_num, timer_idx_t timer_num);
//esp_err_t timer_group_intr_enable(timer_group_t group_num, timer_intr_t intr_mask);
//esp_err_t timer_group_intr_disable(timer_group_t group_num, timer_intr_t intr_mask);


typedef void (*c3_timer_callback)  (void *);

class ESP32FastTimerInterrupt
{
  private:
  
    timer_config_t stdConfig = 
    {
      .alarm_en     = TIMER_ALARM_EN,       //enable timer alarm
      .counter_en   = TIMER_START,          //starts counting counter once timer_init called
      .intr_type    = TIMER_INTR_MAX,
      .counter_dir  = TIMER_COUNT_UP,       //counts from 0 to counter value
      .auto_reload  = TIMER_AUTORELOAD_EN,  // reloads counter automatically
      .divider      = TIMER_DIVIDER
    };

    timer_idx_t       _timerIndex;
    timer_group_t     _timerGroup;
    uint32_t          interruptFlag;        // either TIMER_INTR_T0 or TIMER_INTR_T1
    
    uint8_t           _timerNo;

    c3_timer_callback _callback;        // pointer to the callback function
    float             _frequency;       // Timer frequency
    uint64_t          _timerCount;      // count to activate timer

  public:

    ESP32FastTimerInterrupt(uint8_t timerNo)
    {       
      if (timerNo < MAX_ESP32_NUM_TIMERS)
      {
        _timerNo  = timerNo;
                
        // Always using TIMER_INTR_T0
        _timerIndex = (timer_idx_t)   ( (uint32_t) 0 );
        
        // timerNo == 0 => Group 0, timerNo == 1 => Group 1
        _timerGroup = (timer_group_t) ( (uint32_t) timerNo);
      }
      else
      {
        _timerNo  = MAX_ESP32_NUM_TIMERS;
      }
      
      _frequency  = 0;
      _timerCount = 0;
      _callback = NULL;      
    };

    // frequency (in hertz)
    // No params and duration now. To be added in the future by adding similar functions here or to esp32-hal-timer.c
    bool setFrequency(float frequency, c3_timer_callback callback)
    {
      if (_timerNo < MAX_ESP32_NUM_TIMERS)
      {      
        // select timer frequency is 1MHz for better accuracy. We don't use 16-bit prescaler for now.
        // Will use later if very low frequency is needed.
        _frequency  = TIMER_BASE_CLK / TIMER_DIVIDER;   //1000000;
        _timerCount = (uint64_t) _frequency / frequency;
        // count up
                
        ISR_SERVO_LOGERROR3(F("ESP32_C3_FastTimerInterrupt: _timerNo ="), _timerNo, F(", _fre ="), TIMER_BASE_CLK / TIMER_DIVIDER);
        ISR_SERVO_LOGERROR3(F("TIMER_BASE_CLK ="), TIMER_BASE_CLK, F(", TIMER_DIVIDER ="), TIMER_DIVIDER);
        //ISR_SERVO_LOGERROR3(F("_timerIndex ="), _timerIndex, F(", _timerGroup ="), _timerGroup);
        //ISR_SERVO_LOGERROR3(F("_count ="), (uint32_t) (_timerCount >> 32) , F("-"), (uint32_t) (_timerCount));

        timer_init(_timerGroup, _timerIndex, &stdConfig);
        
        // Counter value to 0 => counting up to alarm value as .counter_dir == TIMER_COUNT_UP
        timer_set_counter_value(_timerGroup, _timerIndex , 0x00000000ULL);
               
        timer_set_alarm_value(_timerGroup, _timerIndex, TIMER_SCALE / frequency);
               
        // enable interrupts for _timerGroup, _timerIndex
        timer_enable_intr(_timerGroup, _timerIndex);
        
        _callback = callback;
        
        // Register the ISR handler       
        // If the intr_alloc_flags value ESP_INTR_FLAG_IRAM is set, the handler function must be declared with IRAM_ATTR attribute
        // and can only call functions in IRAM or ROM. It cannot call other timer APIs.
        timer_isr_register(_timerGroup, _timerIndex, _callback, (void *) (uint32_t) _timerNo, ESP_INTR_FLAG_IRAM, NULL);
        
        return true;
      }
      else
      {
        ISR_SERVO_LOGERROR(F("Error. Timer must be 0-3"));
        
        return false;
      }
    }

    // interval (in microseconds) and duration (in milliseconds). Duration = 0 or not specified => run indefinitely
    // No params and duration now. To be addes in the future by adding similar functions here or to esp32-hal-timer.c
    bool attachInterruptInterval(unsigned long interval, c3_timer_callback callback)
    {
      return setFrequency( (float) ( 1000000.0f / interval), callback);
    }

    void detachInterrupt()
    {
      timer_group_intr_disable(_timerGroup, TIMER_INTR_T0);
    }

    // Duration (in milliseconds). Duration = 0 or not specified => run indefinitely
    void reattachInterrupt()
    {
      if ( (_frequency != 0) && (_timerCount != 0) && (_callback != NULL) )
        timer_group_intr_enable(_timerGroup, TIMER_INTR_T0);
    }
}; // class ESP32FastTimerInterrupt


#endif      // ESP32_C3_FastTimerInterrupt_h
